crypto/evp/keymgmt_lib.c

```diff
@@ -183,6 +200,124 @@ void *evp_keymgmt_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
     return provdata;
 }
 
+int evp_keymgmt_is(const EVP_PKEY *pkey, int domainparams)
+{
+    EVP_KEYMGMT *keymgmt = pkey->pkeys[0].keymgmt;
+
+    if (keymgmt == NULL)
+        return 0;          /* There's no one to ask, so assume "no" */
+    return pkey->pkeys[0].domainparams
+        ? evp_keymgmt_isdomparams(keymgmt, pkey->pkeys[0].provdata)
+        : evp_keymgmt_iskey(keymgmt, pkey->pkeys[0].provdata);
+}
+
+/*
+ * This function adheres to EVP_PKEY_cmp() / EVP_PKEY_cmp_parameters()
+ * return values, i.e.:
+ *
+ *  1   same key
+ *  0   not same key
+ * -1   not same key type
+ * -2   unsupported operation
+ */
+int evp_keymgmt_cmp(const EVP_PKEY *pkey1, const EVP_PKEY *pkey2,
+                    int domainparams)
+{
+    EVP_KEYMGMT *keymgmt1 = pkey1->pkeys[0].keymgmt;
+    EVP_KEYMGMT *keymgmt2 = pkey2->pkeys[0].keymgmt;
+    void *provdata1 = NULL;
+    void *provdata2 = NULL;
+    void *exported_provdata2 = NULL;
+    int rv = 0;                  /* Assume the worst */
+
+    if (keymgmt1 == NULL && keymgmt2 == NULL)
+        return -2;   /* No support without at least one key manager */
+
+    if (keymgmt1 == NULL) {
+        EVP_KEYMGMT *swaptmp = keymgmt1;
+        keymgmt1 = keymgmt2;
+        keymgmt2 = swaptmp;
```

Comment:
Why this elaborate way of assigning NULL to `keymgmt2`?

Reply:
Good question.  Simplifying.